src/services/organizeImports.ts

```diff
@@ -47,8 +55,73 @@ namespace ts.OrganizeImports {
         return changeTracker.getChanges();
     }
 
-    function removeUnusedImports(oldImports: ReadonlyArray<ImportDeclaration>) {
-        return oldImports; // TODO (https://github.com/Microsoft/TypeScript/issues/10020)
+    function removeUnusedImports(oldImports: ReadonlyArray<ImportDeclaration>, sourceFile: SourceFile, program: Program) {
+        const typeChecker = program.getTypeChecker();
+        const jsxNamespace = typeChecker.getJsxNamespace();
+        const jsxContext = sourceFile.languageVariant === LanguageVariant.JSX && program.getCompilerOptions().jsx;
+
+        const usedImports: ImportDeclaration[] = [];
+
+        for (const importDecl of oldImports) {
+            const {importClause} = importDecl;
+
+            if (!importClause) {
+                // Imports without import clauses are assumed to be included for their side effects and are not removed.
+                usedImports.push(importDecl);
+                continue;
+            }
+
+            let { name, namedBindings } = importClause;
+
+            // Default import
+            if (name && !isDeclarationUsed(name)) {
+                name = undefined;
+            }
+
+            if (namedBindings) {
+                if (isNamespaceImport(namedBindings)) {
+                    // Namespace import
+                    if (!isDeclarationUsed(namedBindings.name)) {
+                        namedBindings = undefined;
+                    }
+                }
+                else {
+                    // List of named imports
+                    const newElements = namedBindings.elements.filter(e => isDeclarationUsed(e.propertyName || e.name));
+                    if (newElements.length < namedBindings.elements.length) {
+                        namedBindings = newElements.length
+                            ? updateNamedImports(namedBindings, newElements)
+                            : undefined;
+                    }
+                }
+            }
+
+            if (name || namedBindings) {
+                usedImports.push(updateImportDeclarationAndClause(importDecl, name, namedBindings));
+            }
+        }
+
+        return usedImports;
+
+        function isDeclarationUsed(identifier: Identifier) {

```

Comment:
`return !!program.getTypeChecker().getSymbolAtLocation(identifier)!.isReferenced;`, and at top, `program.getSemanticDiagnostics(sourceFile); // Ensure we've checked this file so `isReferenced` is defined on symbols.`. Also must change check `isUse && result && nameNotFoundMessage && noUnusedIdentifiers` in `checker.ts` to not include `&& noUnusedIdentifiers`, so we always set `isReferenced` accurately.

Reply:
I'll defer to @mhegazy, who asked me to do it this way.

Reply:
Any progress towards reaching a consensus on how to identify unused imports?